File: search/fibonacci_search/javascript/fibonacci-search.js

function fibonacciSearch(arr, x, n)
{
	/* Initialize fibonacci numbers */
	let fibMMm2 = 0; // (m-2)'th Fibonacci No.
	let fibMMm1 = 1; // (m-1)'th Fibonacci No.
	let fibM = fibMMm2 + fibMMm1; // m'th Fibonacci

	/* fibM is going to store the smallest Fibonacci
	Number greater than or equal to n */
	while (fibM < n)
	{
		fibMMm2 = fibMMm1;
		fibMMm1 = fibM;
		fibM = fibMMm2 + fibMMm1;
	}

	// Marks the eliminated range from front
	let offset = -1;

	/* while there are elements to be inspected. Note that
	we compare arr[fibMm2] with x. When fibM becomes 1,
	fibMm2 becomes 0 */

	while (fibM > 1)
	{
		// Check if fibMm2 is a valid location
		let i = Math.min(offset + fibMMm2, n-1);

		/* If x is greater than the value at index fibMm2,
		cut the subarray array from offset to i */
		if (arr[i] < x)
		{
			fibM = fibMMm1;
			fibMMm1 = fibMMm2;
			fibMMm2 = fibM - fibMMm1;
			offset = i;
		}

		/* If x is less than the value at index fibMm2,
		cut the subarray after i+1 */
		else if (arr[i] > x)
		{
			fibM = fibMMm2;
			fibMMm1 = fibMMm1 - fibMMm2;
			fibMMm2 = fibM - fibMMm1;
		}

		/* element found. return index */
		else return i;
	}

	/* comparing the last element with x */
	if(fibMMm1 && arr[n-1] == x){
	return n-1
	}

	/*element not found. return -1 */
	return -1;
}

module.exports = fibonacciSearch;

File: search/fibonacci_search/c/fibonacci_search.c
#include<stdio.h>
#include<conio.h>
 /* This function used to find mininum of given number */
 int min(int x, int y)
 {
     return (x<=y)? x : y;
 }
 /* Returns index of x if present,  else returns -1 */
 int fibonaccianSearch(int arr[], int x, int n)
 {
     /* Here fibonacci numbers are initializing*/
     int fbK2 = 0;   // (k-2)'th Fibonacci number
     int fbK1 = 1;   // (k-1)'th Fibonacci number
     int fbK = fbK2 + fbK1;  // k'th Fibonacci
     // Marks the eliminated range from front
     int offset = -1;
     /* fbK is going to store the smallest Fibonacci
        number greater than or equal to n */
     while (fbK < n)
     {
         fbK2 = fbK1;
         fbK1 = fbK;
         fbK  = fbK2 + fbK1;
     }
     while (fbK > 1)
     {
     // Check if fbK2 is a valid location
     int i = min(offset+fbK2, n-1);
     /* If x is greater than the value at index fbK2,
        cut the subarray array from offset to i */
     if (arr[i] < x)
     {
         fbK  = fbK1;
         fbK1 = fbK2;
         fbK2 = fbK - fbK1;
         offset = i;
     }
     /* If x is greater than the value at index fbKk2,
        cut the subarray after i+1  */
     else if (arr[i] > x)
     {
         fbK  = fbK2;
         fbK1 = fbK1 - fbK2;
         fbK2 = fbK - fbK1;
     }
     /* element found. return index */
     else return i;
     }
     /* comparing the last element with x */
     if(fbK1 && arr[offset+1] == x)
     return offset+1;
     /*element not found. return -1 */
     return -1;
 }
 /* main function */
 int main(void)
 {
     int l,i;
     printf("\nEnter the total number of elements:");
     scanf("%d",&l);

     int arr[l];

     printf("Enter elements in array\n");

     for(i=0;i<l;i++)
     {
         scanf("%d",&arr[i]);
     }
     int n = sizeof(arr)/sizeof(arr[0]);
     int x;

     printf("\nEnter element to be searched:");
     scanf("%d",&x);
     printf("Found at index: %d",fibonaccianSearch(arr, x, n));
     getch();
     return 0;
 }
